staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,145 @@
+	// In tests, emulationVersion could be the same as the binary version, or set directly,
+	// which can have "alpha" as pre-release to continue serving expired apis while we clean up the test.
+	emulationVersion atomic.Pointer[version.Version]
+	// minCompatibilityVersion could only contain major and minor versions.
+	minCompatibilityVersion atomic.Pointer[version.Version]
+}
+
+func (m *effectiveVersion) BinaryVersion() *version.Version {
+	return m.binaryVersion.Load()
+}
+
+func (m *effectiveVersion) EmulationVersion() *version.Version {
+	// Emulation version can have "alpha" as pre-release to continue serving expired apis while we clean up the test.
+	// The pre-release should not be accessible to the users.
+	return m.emulationVersion.Load().WithPreRelease(m.BinaryVersion().PreRelease())
+}
+
+func (m *effectiveVersion) MinCompatibilityVersion() *version.Version {
+	return m.minCompatibilityVersion.Load()
+}
+
+func (m *effectiveVersion) EqualTo(other EffectiveVersion) bool {
+	return m.BinaryVersion().EqualTo(other.BinaryVersion()) && m.EmulationVersion().EqualTo(other.EmulationVersion()) && m.MinCompatibilityVersion().EqualTo(other.MinCompatibilityVersion())
+}
+
+func (m *effectiveVersion) String() string {
+	if m == nil {
+		return "<nil>"
+	}
+	return fmt.Sprintf("{BinaryVersion: %s, EmulationVersion: %s, MinCompatibilityVersion: %s}",
+		m.BinaryVersion().String(), m.EmulationVersion().String(), m.MinCompatibilityVersion().String())
+}
+
+func (m *effectiveVersion) Set(binaryVersion, emulationVersion, minCompatibilityVersion *version.Version) {
+	m.binaryVersion.Store(binaryVersion)
+	m.emulationVersion.Store(version.MajorMinor(emulationVersion.Major(), emulationVersion.Minor()))
+	m.minCompatibilityVersion.Store(version.MajorMinor(minCompatibilityVersion.Major(), minCompatibilityVersion.Minor()))
+}
+
+func (m *effectiveVersion) SetEmulationVersion(emulationVersion *version.Version) {
+	m.emulationVersion.Store(version.MajorMinor(emulationVersion.Major(), emulationVersion.Minor()))
+}
+
+func (m *effectiveVersion) SetMinCompatibilityVersion(minCompatibilityVersion *version.Version) {
+	m.minCompatibilityVersion.Store(version.MajorMinor(minCompatibilityVersion.Major(), minCompatibilityVersion.Minor()))
+}
+
+func (m *effectiveVersion) Validate() []error {
+	var errs []error
+	// Validate only checks the major and minor versions.
+	binaryVersion := m.binaryVersion.Load().WithPatch(0)
+	emulationVersion := m.emulationVersion.Load()
+	minCompatibilityVersion := m.minCompatibilityVersion.Load()
+
+	// emulationVersion can only be 1.{binaryMinor-1}...1.{binaryMinor}.
+	maxEmuVer := binaryVersion
+	minEmuVer := binaryVersion.SubtractMinor(1)
+	// TODO: remove in 1.32
+	// emulationVersion is introduced in 1.31, so it cannot be lower than that.
+	// binaryVersion could be lower than 1.31 in tests. So we are only checking 1.31.
+	if binaryVersion.EqualTo(version.MajorMinor(1, 31)) {
+		minEmuVer = version.MajorMinor(1, 31)
+	}
```

Comment:
how does this work when my product is versioned 4.18?  oversight?

Reply:
We don't expect emulation version to be widely adopted in alpha. Clients can always implement their own check separately. 
We had to hard code the 1.31 version instead of forcing emulationVersion=binaryVersion to work around the tests. But we cannot know what versions the client's binary would be.

Reply:
> But we cannot know what versions the client's binary would be.

Agreed.  This is validation is in the wrong spot.  This is kube-specific, not generic to all apiservers.  this code should move to a kube-apiserver specific place.